### DISCUSSION TITLE
Birdseye export? (r13)

### DISCUSSION THREAD

**Discussion by PatSloan** [blakeblackshear/frigate#9272] [Ideas]
I think it might be cool to have a birdseye export.   Thoughts?

**Comment by NickM-27** [comment]
What do you mean by export? Exported in what way?

**Reply by PatSloan** [reply]
oh, in r13 there is an export function.  It allows you to export a given camera feed for a period of time it was recorded.  includes option to create a timelapse output.  pretty cool. 

Guess, you already know that ;)

So i'd just like the ability to do the same, but the output is the equivalent to the birdseye view.

**Reply by NickM-27** [reply]
that would require birdseye to be recorded, which can already be done manually by adding a birdseye camera and making it recorded only, but I am not sure birdseye recording would be offered natively

**Comment by PatSloan** [comment]
> that would require birdseye to be recorded, which can already be done manually by adding a birdseye camera and making it recorded only, but I am not sure birdseye recording would be offered natively

So even if all the streams that make up the birdseye view are recorded, it can't combine them on the export, just like it does for the live stream? 

--pat

**Reply by NickM-27** [reply]
That's not how it works, the birdseye stream is manually combined using raw frames directly from each camera to reduce overhead, it is not stitching the live views together directly. Doing that during an export directly would be an incredible task for a GPU and would likely grind the host device to a halt. It also wouldn't have as much flexibility on how the streams are combined.

**Reply by PatSloan** [reply]
thanks for clarifying.

**Comment by PatSloan** [comment]
> that would require birdseye to be recorded, which can already be done manually by adding a birdseye camera and making it recorded only, but I am not sure birdseye recording would be offered natively

are there any pointers on how to setup birdseye as a camera and record it?

**Reply by NickM-27** [reply]
when you enable birdseye restream it is available at rtsp://127.0.0.1:8554/birdseye so that address can be added as a separate camera that will only have recording enabled

**Reply by heisenberg2980** [reply]
@NickM-27 I setup birdseye as a new camera using the address you mentioned, and it appears as a new camera, but it doesn´t record anything. Looking at the logs it seems the ffmpeg process for the new birdseye camera is crashing:

```
2024-08-09 16:17:06.678761814  [2024-08-09 16:17:06] frigate.video                  ERROR   : birdseye: Unable to read frames from ffmpeg process.
2024-08-09 16:17:06.678882288  [2024-08-09 16:17:06] frigate.video                  ERROR   : birdseye: ffmpeg process is not running. exiting capture thread...
2024-08-09 16:17:16.685517317  [2024-08-09 16:17:16] watchdog.birdseye              ERROR   : Ffmpeg process crashed unexpectedly for birdseye.
2024-08-09 16:17:16.685580985  [2024-08-09 16:17:16] watchdog.birdseye              ERROR   : The following ffmpeg logs include the last 100 lines prior to exit.
2024-08-09 16:17:16.685607414  [2024-08-09 16:17:16] ffmpeg.birdseye.detect         ERROR   : [rtsp @ 0x5660e62382c0] method DESCRIBE failed: 404 Not Found
2024-08-09 16:17:16.685608126  [2024-08-09 16:17:16] ffmpeg.birdseye.detect         ERROR   : rtsp://127.0.0.1:8554/birdseye: Server returned 404 Not Found
```

**Reply by NickM-27** [reply]
Did you setup restream as true?

**Reply by heisenberg2980** [reply]
@NickM-27 Yes I did, this is how birdseye is configured in my config (I have tried commenting out the "mode: continuous" but got the same result):

```
birdseye:
  mode: continuous
  restream: true
  width: 1920
  height: 1080
```

After more analysis I realised there is an error in the logs just after starting Frigate when the birdseye camera is enable, this is the error (File exists: '/birdseye') but I´m not sure what it means:

```
frigate  | 2024-08-09 21:21:25.322777203  Process output_processor:
frigate  | 2024-08-09 21:21:25.324075655  Traceback (most recent call last):
frigate  | 2024-08-09 21:21:25.324077609    File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
frigate  | 2024-08-09 21:21:25.324080905      self.run()
frigate  | 2024-08-09 21:21:25.324082548    File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
frigate  | 2024-08-09 21:21:25.324083520      self._target(*self._args, **self._kwargs)
frigate  | 2024-08-09 21:21:25.324084191    File "/opt/frigate/frigate/output.py", line 816, in output_frames
frigate  | 2024-08-09 21:21:25.324099810      birdseye_buffer = frame_manager.create(
frigate  | 2024-08-09 21:21:25.324100682    File "/opt/frigate/frigate/util/image.py", line 644, in create
frigate  | 2024-08-09 21:21:25.324101673      shm = shared_memory.SharedMemory(name=name, create=True, size=size)
frigate  | 2024-08-09 21:21:25.324102355    File "/usr/lib/python3.9/multiprocessing/shared_memory.py", line 103, in __init__
frigate  | 2024-08-09 21:21:25.324121841      self._fd = _posixshmem.shm_open(
frigate  | 2024-08-09 21:21:25.324122472  FileExistsError: [Errno 17] File exists: '/birdseye'
```

**Reply by NickM-27** [reply]
do you have /dev/shm mounted in docker? it should not be

**Reply by heisenberg2980** [reply]
No I don´t have /dev/shm mounted in docker, but I just realised the reason for the failure. The issue was that I named the new camera "birdseye", and it seems you can´t do that as that name must be used by the original birdseye stream. I renamed the camera to birdseye_cam and it works now. 

The only pending issue was that after creating the new camera a kind of infinite loop was created as the birdseye view included the birdseye_cam camera. For anyone reading this with the same issue, this can be easily solved by disabling birdseye in the new birdseye_cam, so this camera is not included in the birdseye view. This is my final configuration for the birdseye_cam camera:

```
  birdseye_cam: 
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/birdseye 
          roles:
            - record
    record: 
      enabled: True
    audio: 
      enabled: false
    birdseye:
      enabled: False
```

Thank you very much @NickM-27, you guided me to find the solution to this problem.
